crates/nu-command/src/filters/group_by.rs

```diff
@@ -127,6 +127,7 @@ pub fn group_by(
     let head = call.head;
     let grouper: Option<Value> = call.opt(engine_state, stack, 0)?;
     let to_table = call.has_flag(engine_state, stack, "to-table")?;
+    let config = engine_state.get_config();
```

Comment:
That the `group-by` results can change depending on the user config sounds like bad design.

If you have a string around that matches the one particular potential formatting based on the config, this could also change the group assignment, but at the very least the columns of the resulting record if a value that has configurable formatting rules.

In general we run into two limitations of the current state of the type system/internal representation.
One we don't easily change even from a user perspective is that our records have string keys and there is no equivalent `Value -> Value` map type we could use not requiring a string representation and having problems when multiple values collide in the string space.
The second one is also gnarly is defining a `Hash`/`Eq` implementation pair for `Value` so operations like grouping or uniq could happen on the `Value` level. There the challenge is partially driven by floats and choices like should equivalent floats and ints hash the same if they are numerically equal. Anything there will be a compromise.

Reply:
In the Discord discussion about this topic, I was complaining about the many functions we have to convert from value to string and not knowing when each one is appropriate when all I want is a string. @IanManske suggested using `to_abbreviated_string()` instead of adding more match arms to coerce_str/string, which is where the problem is right now with grouping by bools without this PR. The config is used because we allow datetime and filesize formats to be changed via config.

I also looked at just using a Value for grouping but quickly stumbled upon the Hash problem you mentioned above.

The other option is to have yet another value to string function that just uses a consistent format for datetime and filesize regardless of config, but I'm sure people would complain. I'd be ok with that as long as we get rid of some of the other value to string functions.

Reply:
In the current signature for `group-by` we will have to accept some level of weird. I think we should clearly state what weirdness we are OK with and the users have to be aware of. 
This could mean:
- some types are not supported and you need to prepare the values specifically yourself based on the semantics the user wants.
- `true` and `"true"` will cluster into the same bucket or alternatively cause an error.
- User config changes stuff for the formattable values or there is a canonical config with which things are string keyed. Then if you as a user want to retrieve based on one e.g. `datetime` value you need to convert that value using the same universal default formatting to a string to retrieve the group-by bucket.
- (making sure that there is nothing in the string conversion that is dependent on factors outside the (static) config. i.e. relative time info on showing datetime, localization env vars.)

Reply:
Ya, I agree. This is less about coding and more about deciding what we want for nushell and then putting it in the `extra_description()`

Reply:
Thanks! The text is great in my view. A lot more of our commands should have the semantics written on the packaging.

I am still not a fan of the config dependence on the other hand maybe we should make it available and explicitly elicit user feedback on that. I don't have a clear idea what the behavior should be for all the types, and maybe we gain more by collecting several specific complaints and after we get the feedback can make one change to a more static solution. (having python `hash()` flashbacks right now :D )